Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,266 @@
+* TailCallHelperStack - a per thread stack of helper entries that is used to determine whether a tail call is chained or not. Its entries are allocated as local variables in CallTarget thunks. Each entry contains:
+  * Stack pointer captured right before a call to a tail call target
+  * ChainCall flag indicating whether the CallTarget thunk should return after the call to the tail call target or whether it should execute its epilog and jump to TailCallHelper instead. The latter is used by the TailCallHelper to remove the stack frame of the CallTarget before making a tail call from a tail called function. 
+  * Pointer to the next entry on the stack.
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
+* The caller executes its epilog, restoring stack pointer and callee saved registers to their values when the caller was entered.
+* The caller jumps to the TailCallHelper. This function performs the following operations:
+  * Get the topmost TailCallHelperStack entry for the current thread.
+  * Check if the previous stack frame is a CallTarget thunk frame by comparing the stack pointer value stored in the TailCallHelperStack entry to the current CFA (call frame address). If it matches, it means that the previous stack frame belongs to a CallTarget thunk and so the tail call caller was also tail called.
+  * If the previous frame was a CallTarget thunk, its stack frame needs to be removed to ensure that the stack will not grow when tail calls are chained. Set ChainCall flag in the TailCallHelperStack entry and return. That returns control to the CallTarget thunk. It checks the ChainCall flag and since it set, it executes its epilog and jumps to the TailCallHelper again.
+  * If the previous frame was not a CallTarget thunk, get the address of the CallTarget thunk of the tailcall target from the arguments buffer and jump to it.
+* The CallTarget thunk function then does the following operation:
+  * Create local instance of TailCallHelperStack entry and store the current stack pointer value in it.
+  * Push the entry to the TailCallHelperStack of the current thread.
+  * Get the arguments buffer from the thread local storage, extract the arguments, release the buffer and call the target function. The frame of the CallTarget thunk ensures that the arguments of the target are GC protected until the target function returns or tail calls to another function.
+  * Pop the TailCallHelperStack entry from the TailCallHelperStack of the current thread.
+  * Check the ChainCall flag in the TailCallHelperStack entry. If it is set, run epilog and jump to the TailCallHelper.
+  * If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function.
```

Comment:
@sivarv one of the cases where the generic context cannot be the same for CallTarget and the tailcall target is the case of a generic method in a non-generic class. In that case, the generic context is the MethodDesc of the method and you cannot have two different methods having the same MethodDesc. 
So we really need to pass the generic context of the CallTarget and not the one of the tailcall target to the TailCallHelper. 
As for your question:
> Say we are jit compiling B(). At this time JIT knows the generic context that needs to be passed to C(). How will JIT know the generic context that needs to be passed to CallTarget IL thunk?

I don't know how the generic context is figured out in the JIT, but since it can figure it out for a call to C, couldn't it use the same way to figure it out for the CallTarget?
Actually, what if we did things the way that the jitter would generate code as if it was jumping to the CallTarget (but jump to the TailCallHelper instead)? It seems that that would ensure that it would pass in the proper generic context of the CallTarget in a natural way.